File: tablo_olustur.R
options(repos = c(CRAN = "https://cloud.r-project.org"))

# Gerekli paketler
if (!require("readr")) install.packages("readr")
if (!require("dplyr")) install.packages("dplyr")
if (!require("broom")) install.packages("broom")
if (!require("epitools")) install.packages("epitools")
if (!require("writexl")) install.packages("writexl")

library(readr)
library(dplyr)
library(broom)
library(epitools)
library(writexl)

# Veriyi oku
df <- read_delim("Test.csv", delim = ";", locale = locale(decimal_mark = ",", grouping_mark = "."))

# Değişken listesi (örnek, ihtiyaca göre genişletilebilir)
variables <- c(
  "Age", "Sex", "BMI", "Type", "Grade (CNB)", "Histology (CNB)", "Weight", "Hypertension", "Smoker",
  "Previous surgery", "Diabetes", "Grade (P)", "Type (P)", "Histology (P)", "BRCA ", "pT", "pN",
  "N(x/x)" = "N(X/x)", "N(X/x)", "N(x/X)" = "N (x/X)", "M", "L", "V", "Pn", "CRP", "Temp",
  "Duration", "Size", "Mastectomy", "Axilla", "CTX", "Axilla 24h", "Mamma 24h", "Drainage (A)",
  "Drainage  (M)", "Opiod", "Lenght of stay", "Follow-Up", "Total (A)", "Total (M)", "Total"
)

# Sonuç tablosu için boş bir data.frame
results <- data.frame(
  Variable = character(),
  N = integer(),
  Min = numeric(),
  Max = numeric(),
  Mean_SD = character(),
  Median = numeric(),
  Std_Dev = numeric(),
  Range = character(),
  OR_95CI = character(),
  P_value = character(),
  Comparison = character(),
  stringsAsFactors = FALSE
)

# Mastectomy'yi faktör olarak ayarla
df$Mastectomy <- as.factor(df$Mastectomy)

for (var in variables) {
  if (!(var %in% names(df))) {
    results <- rbind(results, data.frame(
      Variable = var, N = NA, Min = NA, Max = NA, Mean_SD = NA, Median = NA, Std_Dev = NA, Range = NA, OR_95CI = NA, P_value = NA, Comparison = NA
    ))
    next
  }
  x <- df[[var]]
  Comparison <- ""
  # Sayısal mı kategorik mi?
  if (is.numeric(x) || is.integer(x)) {
    N <- sum(!is.na(x))
    Min <- min(x, na.rm = TRUE)
    Max <- max(x, na.rm = TRUE)
    Mean <- mean(x, na.rm = TRUE)
    SD <- sd(x, na.rm = TRUE)
    Median <- median(x, na.rm = TRUE)
    Range <- paste0(Min, "-", Max)
    Mean_SD <- sprintf("%.2f ± %.2f", Mean, SD)
    Std_Dev <- sprintf("%.3f", SD)
    # Mastectomy'ye göre karşılaştırma (t-testi veya Mann-Whitney)
    if (length(unique(na.omit(df$Mastectomy))) == 2) {
      g0 <- x[df$Mastectomy == levels(df$Mastectomy)[1]]
      g1 <- x[df$Mastectomy == levels(df$Mastectomy)[2]]
      if (length(g0) > 1 && length(g1) > 1) {
        pval <- tryCatch(
          t.test(g0, g1)$p.value,
          error = function(e) tryCatch(
            wilcox.test(g0, g1)$p.value,
            error = function(e) NA
          )
        )
        P_value <- sprintf("%.3g", pval)
        Comparison <- paste0("Mastectomy: ", levels(df$Mastectomy)[1], " vs ", levels(df$Mastectomy)[2])
      } else {
        P_value <- NA
      }
    } else {
      P_value <- NA
    }
    OR_95CI <- ""
  } else if (is.factor(x) || is.character(x) || length(unique(na.omit(x))) <= 10) {
    # Kategorik değişkenler için
    tab <- table(x, df$Mastectomy)
    N <- sum(!is.na(x))
    Min <- Max <- Mean_SD <- Median <- Std_Dev <- Range <- NA
    # Sadece 2x2 tablo için OR ve p-value
    if (all(dim(tab) == 2) && all(tab > 0)) {
      or_res <- oddsratio(tab)
      OR <- or_res$measure[2,1]
      CI_low <- or_res$measure[2,2]
      CI_high <- or_res$measure[2,3]
      OR_95CI <- sprintf("%.2f (%.2f-%.2f)", OR, CI_low, CI_high)
      P_value <- sprintf("%.3g", fisher.test(tab)$p.value)
      Comparison <- paste0("Mastectomy: ", colnames(tab)[1], " vs ", colnames(tab)[2])
    } else {
      OR_95CI <- ""
      P_value <- ""
      Comparison <- ""
    }
  } else {
    N <- Min <- Max <- Mean_SD <- Median <- Std_Dev <- Range <- OR_95CI <- P_value <- Comparison <- NA
  }
  results <- rbind(results, data.frame(
    Variable = var, N = N, Min = Min, Max = Max, Mean_SD = Mean_SD, Median = Median, Std_Dev = Std_Dev, Range = Range, OR_95CI = OR_95CI, P_value = P_value, Comparison = Comparison
  ))
}

# Seroma için boş satır ekle
results <- rbind(
  results,
  data.frame(Variable = "Seroma", N = NA, Min = NA, Max = NA, Mean_SD = NA, Median = NA, Std_Dev = NA, Range = NA, OR_95CI = NA, P_value = NA, Comparison = NA)
)

# --- Size için p-value ---
size_pvalue <- ""
if ("Size" %in% names(df) && length(unique(na.omit(df$Mastectomy))) == 2) {
  size <- df$Size
  g0 <- size[df$Mastectomy == levels(df$Mastectomy)[1]]
  g1 <- size[df$Mastectomy == levels(df$Mastectomy)[2]]
  if (length(g0) > 1 && length(g1) > 1) {
    size_pval <- tryCatch(
      t.test(g0, g1)$p.value,
      error = function(e) tryCatch(
        wilcox.test(g0, g1)$p.value,
        error = function(e) NA
      )
    )
    size_pvalue <- as.character(signif(size_pval, 3))
  }
}
# Tabloya Size'ın p-value'sini karakter olarak ekle
results$P_value[results$Variable == "Size"] <- as.character(size_pvalue)

# --- Mastectomy için istatistikler ---
mastectomy <- as.numeric(as.character(df$Mastectomy))
N <- sum(!is.na(mastectomy))
Min <- min(mastectomy, na.rm = TRUE)
Max <- max(mastectomy, na.rm = TRUE)
Mean <- mean(mastectomy, na.rm = TRUE)
SD <- sd(mastectomy, na.rm = TRUE)
Median <- median(mastectomy, na.rm = TRUE)
Range <- paste0(Min, "-", Max)
Mean_SD <- sprintf("%.2f ± %.2f", Mean, SD)
Std_Dev <- sprintf("%.3f", SD)
P_value <- NA
Comparison <- ""
if (length(unique(na.omit(mastectomy))) == 2) {
  g0 <- mastectomy[mastectomy == 0]
  g1 <- mastectomy[mastectomy == 1]
  if (length(g0) > 1 && length(g1) > 1) {
    pval <- tryCatch(
      t.test(g0, g1)$p.value,
      error = function(e) tryCatch(
        wilcox.test(g0, g1)$p.value,
        error = function(e) NA
      )
    )
    P_value <- sprintf("%.3g", pval)
    Comparison <- "Mastectomy: 0 vs 1"
  }
}
mastectomy_row <- data.frame(
  Variable = "Mastectomy (summary)",
  N = as.character(N),
  Min = as.character(Min),
  Max = as.character(Max),
  Mean_SD = as.character(Mean_SD),
  Median = as.character(Median),
  Std_Dev = as.character(Std_Dev),
  Range = as.character(Range),
  OR_95CI = "",
  P_value = as.character(P_value),
  Comparison = as.character(Comparison),
  stringsAsFactors = FALSE
)
mastectomy_row[] <- lapply(mastectomy_row, as.character)
results2 <- rbind(results, mastectomy_row)
results2[] <- lapply(results2, as.character)
write_xlsx(results2, "Tablo_Sonuclari_R.xlsx")
cat("Tablo başarıyla oluşturuldu ve Tablo_Sonuclari_R.xlsx olarak kaydedildi.\n") 

File: tablo_olustur.py
import pandas as pd
import numpy as np
from scipy.stats import chi2_contingency, fisher_exact, ttest_ind, mannwhitneyu
from statsmodels.stats.contingency_tables import Table2x2

# Dosyayı oku
# Not: encoding ve delimiter ayarları gerekirse değiştirilebilir
df = pd.read_csv("Test.csv", sep=";", encoding="latin1")

# Görseldeki tabloya göre başlıkları csv ile eşleştir
variables = [
    # Epidemiological
    ("Age", "Age"),
    ("Sex", "Sex"),
    ("BMI", "BMI"),
    # Clinical
    ("Type", "Type"),
    ("Grade (CNB)", "Grade (CNB)"),
    ("Histology (CNB)", "Histology (CNB)"),
    ("Weight", "Weight"),
    ("Hypertension", "Hypertension"),
    ("Smoker", "Smoker"),
    ("Previous surgery", "Previous surgery"),
    ("Diabetes", "Diabetes"),
    ("Grade (P)", "Grade (P)"),
    ("Type (P)", "Type (P)"),
    ("Histology (P)", "Histology (P)"),
    ("BRCA", "BRCA "),
    ("pT", "pT"),
    ("pN", "pN"),
    ("N(x/x)", "N(X/x)"),
    ("N(X/x)", "N(X/x)"),
    ("N(x/X)", "N (x/X)"),
    ("M", "M"),
    ("L", "L"),
    ("V", "V"),
    ("Pn", "Pn"),
    ("CRP", "CRP"),
    ("Temp", "Temp"),
    # Therapeutic
    ("Duration", "Duration"),
    ("Size", "Size"),
    ("Mastectomy", "Mastectomy"),
    ("Axilla", "Axilla"),
    ("CTX", "CTX"),
    ("Axilla 24h", "Axilla 24h"),
    ("Mamma 24h", "Mamma 24h"),
    ("Drainage (A)", "Drainage (A)"),
    ("Drainage (M)", "Drainage  (M)"),
    ("Opioid", "Opiod"),
    ("Length of stay", "Lenght of stay"),
    ("Seroma", None),
    # Seroma ve altı
    ("Follow-Up", "Follow-Up"),
    ("Total (A)", "Total (A)"),
    ("Total (M)", "Total (M)"),
    ("Total", "Total"),
]

results = []

# Mastectomy iki grup: 0 ve 1 (veya varsa diğer değerler)
mastectomy_col = "Mastectomy"
if mastectomy_col in df.columns:
    group0 = df[df[mastectomy_col] == 0]
    group1 = df[df[mastectomy_col] == 1]
else:
    group0 = group1 = pd.DataFrame()

for label, col in variables:
    if col and col in df.columns:
        data = pd.to_numeric(df[col], errors='coerce')
        N = data.count()
        Min = data.min()
        Max = data.max()
        Mean_SD = f"{data.mean():.2f} ± {data.std():.2f}" if N > 0 else ""
        Median = data.median()
        Std_Dev = data.std()
        Range = f"{Min}-{Max}" if N > 0 else ""
        OR_CI = ""
        P_value = ""
        # Kategorik mi sürekli mi?
        unique_vals = df[col].dropna().unique()
        if len(unique_vals) <= 10 and df[col].dropna().dtype in [np.int64, np.float64, int, float]:
            # Kategorik değişken (ör: 0/1)
            try:
                tab = pd.crosstab(df[col], df[mastectomy_col])
                if tab.shape == (2,2):
                    # OR ve %95 CI
                    table = Table2x2(tab.values)
                    OR = table.oddsratio
                    CI_low, CI_upp = table.oddsratio_confint()
                    OR_CI = f"{OR:.2f} ({CI_low:.2f}-{CI_upp:.2f})"
                    # Fisher exact veya ki-kare
                    _, p = fisher_exact(tab.values)
                    P_value = f"{p:.4f}"
                else:
                    P_value = "-"
            except Exception:
                P_value = "-"
        elif group0.shape[0] > 0 and group1.shape[0] > 0:
            # Sürekli değişkenlerde iki grup karşılaştırması
            g0 = pd.to_numeric(group0[col], errors='coerce').dropna()
            g1 = pd.to_numeric(group1[col], errors='coerce').dropna()
            if len(g0) > 0 and len(g1) > 0:
                # Normal dağılım varsayımı için t-testi, değilse Mann-Whitney U
                try:
                    stat, p = ttest_ind(g0, g1, nan_policy='omit')
                except Exception:
                    stat, p = mannwhitneyu(g0, g1, alternative='two-sided')
                P_value = f"{p:.4f}"
        results.append([label, N, Min, Max, Mean_SD, Median, Std_Dev, Range, OR_CI, P_value])
    else:
        results.append([label, "", "", "", "", "", "", "", "", ""])

columns = ["Variable", "N", "Min", "Max", "Mean_SD", "Median", "Std_Dev", "Range", "OR_95CI", "P_value"]
result_df = pd.DataFrame(results, columns=columns)
result_df.to_excel("Tablo_Sonuclari.xlsx", index=False)
print("Tablo başarıyla oluşturuldu ve Tablo_Sonuclari.xlsx olarak kaydedildi.") 